### DISCUSSION TITLE
Help Me plz

### DISCUSSION THREAD

**Discussion by Arshkamran1** [deta/cloud-docs#723] [Help]
My code uses google sheets, and then sends an email to the last row added. It runs locally but when I run it in the micro I get this Running micro...

Response:
{
        "error": "No function found for action 'an empty action'"
} I have installed all the dependencies what should I do

**Comment by jnsougata** [comment]
Can you share some code we can look at?

**Reply by Arshkamran1** [reply]
here is the code for main.py 
from googleapiclient.discovery import build
import pandas as pd
import time
from send_email import send_email
from google.oauth2.credentials import Credentials
from google.oauth2.service_account import Credentials
import json
from deta import app


@app.lib.cron()
def cron_job(event):

    spreadsheet_id = '1422gMeYwrhGoMXasvQ8_FDoxiTAblM5yivfJyYO3w7M'

    range_name = 'Sheet1!A1:Z1000'

    creds = Credentials.from_service_account_file(
        'email-374509-4f72425b93d2.json', scopes=["https://www.googleapis.com/auth/spreadsheets"])

    service = build('sheets', 'v4', credentials=creds)

    result = service.spreadsheets().values().get(
        spreadsheetId=spreadsheet_id, range=range_name).execute()

    df = pd.DataFrame(result.get('values', []))

    initial_rows = df.shape[0]

    while True:

        result = service.spreadsheets().values().get(
            spreadsheetId=spreadsheet_id, range=range_name).execute()
        df = pd.DataFrame(result.get('values', []))

        current_rows = df.shape[0]
        if current_rows > initial_rows:

            send_email(df.iloc[-1])

            initial_rows = current_rows

        time.sleep(60)

and here is the code for send_email function

import os
import smtplib
from email.message import EmailMessage
from email.utils import formataddr
from pathlib import Path
from dotenv import load_dotenv


PORT = 587
EMAIL_SERVER = "smtp.gmail.com"

current_dir = Path(__file__).resolver(
).parent if "_file_" in locals() else Path.cwd()
envars = current_dir / "email.env"
load_dotenv(envars)

sender_email = os.getenv("EMAIL")
password_email = os.getenv("PASSWORD")
app_password = os.getenv("APP_PASSWORD")


def send_email(row):
    subject = "Your order is on its way"
    name = row[0]
    last_name = row[1]
    reciever_email = row[2]
    number = row[4]
    street_address = row[5]
    city = row[6]
    province = row[7]
    postal_code = row[8]
    payment_method = row[9]
    date = row[10]
    items = row[13]

    with open('email_template.html', 'r') as file:
        email_template = file.read()
    email_template = email_template.format(name=name, last_name=last_name, city=city,
                                           province=province, postal_code=postal_code, street_address=street_address, number=number, reciever_email=reciever_email)

    msg = EmailMessage()
    msg['Subject'] = subject
    msg["From"] = formataddr(("Khana", f"{sender_email}"))
    msg["To"] = row[0]

    msg.set_content(
        f"""\
            First Name: {name},
            City: {city}
            Province: {province}
            Postal Code: {postal_code}
            Street Address: {street_address}
            Number: {number}
            Last Name: {last_name}
            """

    )

    msg.add_alternative(
        f"""\<html>
<head>
  <title>Order Confirmation</title>
</head>
<body>
  <h1>Your order has been placed</h1>
  <h2> Below our the details of your order</h2>
  <p>Date:{date}</p>
  <p>First Name:{name}</p>
  <p>Last Name:{last_name}</p>
  <p>Email Address:{reciever_email}</p>
  <p>Phone Number:{number}</p>
  <p>Street Address:{street_address}</p>
  <p>City:{city}</p>
  <p>Province:{province}</p>
  <p>Postal Code:{postal_code}</p>
  <p>Payment Method:{payment_method}</p>
  <p>Order Details:{items}</p>
  <p>This is a test</p>
</body>
</html>
""",
subtype="html"
    )

**Reply by Arshkamran1** [reply]
srry this is the code for send_email function

import os
import smtplib
from email.message import EmailMessage
from email.utils import formataddr
from pathlib import Path
from dotenv import load_dotenv


PORT = 587
EMAIL_SERVER = "smtp.gmail.com"

current_dir = Path(__file__).resolver(
).parent if "_file_" in locals() else Path.cwd()
envars = current_dir / "email.env"
load_dotenv(envars)

sender_email = os.getenv("EMAIL")
password_email = os.getenv("PASSWORD")
app_password = os.getenv("APP_PASSWORD")


def send_email(row):
    subject = "Your order is on its way"
    name = row[0]
    last_name = row[1]
    reciever_email = row[2]
    number = row[4]
    street_address = row[5]
    city = row[6]
    province = row[7]
    postal_code = row[8]
    payment_method = row[9]
    date = row[10]
    items = row[13]

    with open('email_template.html', 'r') as file:
        email_template = file.read()
    email_template = email_template.format(name=name, last_name=last_name, city=city,
                                           province=province, postal_code=postal_code, street_address=street_address, number=number, reciever_email=reciever_email)

    msg = EmailMessage()
    msg['Subject'] = subject
    msg["From"] = formataddr(("Khana", f"{sender_email}"))
    msg["To"] = row[0]

    msg.set_content(
        f"""\
            First Name: {name},
            City: {city}
            Province: {province}
            Postal Code: {postal_code}
            Street Address: {street_address}
            Number: {number}
            Last Name: {last_name}
            """

    )

    msg.add_alternative(
        f"""\<html>
<head>
  <title>Order Confirmation</title>
</head>
<body>
  <h1>Your order has been placed</h1>
  <h2> Below our the details of your order</h2>
  <p>Date:{date}</p>
  <p>First Name:{name}</p>
  <p>Last Name:{last_name}</p>
  <p>Email Address:{reciever_email}</p>
  <p>Phone Number:{number}</p>
  <p>Street Address:{street_address}</p>
  <p>City:{city}</p>
  <p>Province:{province}</p>
  <p>Postal Code:{postal_code}</p>
  <p>Payment Method:{payment_method}</p>
  <p>Order Details:{items}</p>
  <p>This is a test</p>
</body>
</html>
""",
subtype="html"
    )
        

    with smtplib.SMTP(EMAIL_SERVER, PORT) as server:
        server.starttls()
        server.login(sender_email, app_password)
        server.sendmail(sender_email, reciever_email, msg.as_string())


if __name__ == "__main__":
    send_email(row=["bob", "bob",
               "bob@gmail.com", "Waterloo", "647", "horizon", "", "Ontario", "L4B"])

**Reply by jnsougata** [reply]
Sorry for replying late. Are you getting any specific errors?
